client/src/utils/ajax.ts

```diff
@@ -0,0 +1,223 @@
+import envData from '../../../config/env.json';
+import Tokens from 'csrf';
+import cookies from 'browser-cookies';
+
+const { apiLocation } = envData as { apiLocation: string };
+
+const base = apiLocation;
+const tokens = new Tokens();
+
+const defaultOptions: RequestInit = {
+  credentials: 'include'
+};
+
+// _csrf is passed to the client as a cookie. Tokens are sent back to the server
+// via headers:
+function getCSRFToken() {
+  const _csrf = typeof window !== 'undefined' && cookies.get('_csrf');
+  if (!_csrf) {
+    return '';
+  } else {
+    return tokens.create(_csrf);
+  }
+}
+
+async function get<T>(path: string): Promise<T> {
+  const res = await fetch(`${base}${path}`, defaultOptions);
+  return (await res.json()) as Promise<T>;
+}
+
+export function post<T = void>(path: string, body: unknown): Promise<T> {
+  return request('POST', path, body);
+}
+
+function put<T = void>(path: string, body: unknown): Promise<T> {
+  return request('PUT', path, body);
+}
+
+async function request<T>(
+  method: 'POST' | 'PUT',
+  path: string,
+  body: unknown
+): Promise<T> {
+  const options: RequestInit = {
+    ...defaultOptions,
+    method,
+    headers: {
+      'CSRF-Token': getCSRFToken(),
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify(body)
+  };
+  const res = await fetch(`${base}${path}`, options);
+  return (await res.json()) as Promise<T>;
+}
+
+/** GET **/
+
+interface ProfileUI {
```

Comment:
Most, if not all, of these types are defined in `client/src/redux/prop-types`. We should probably take the opportunity to move this file, but probably not in this PR. So, these types should just be imported until then.